ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+      let kbInfo = request.userInfo[Key.keyboardInfo]
+      let keyboard: [String: String]
+      if let kb = kbInfo as? [String: String] {
+        keyboard = kb
+      } else if let keyboards = kbInfo as? [[String: String]], !keyboards.isEmpty {
+        keyboard = keyboards[0]
+      } else {
+        kmLog("downloadRequestFailed(): Unexpected userInfo: \(String(describing: kbInfo))",
+          checkDebugPrinting: false)
+        return
+      }
+      let kbID = keyboard[Key.keyboardId]!
+      let kbVersion = keyboard[Key.keyboardVersion]
+      let isUpdate = (request.userInfo[Key.update] as? Int)! != 0
+
+      if !isUpdate {
+        let fileManager = FileManager.default
+        let fileName = request.url.lastPathComponent
+        if fileName.hasJavaScriptExtension {
+          if let kbPath = keyboardPath(forFilename: fileName, keyboardVersion: kbVersion) {
+            try? fileManager.removeItem(at: kbPath)
+          }
+        }
+        if fileName.hasFontExtension {
+          // TODO: Check why this doesn't match the error case in downloadRequestFinished().
+          if let kbPath = keyboardPath(forID: kbID, keyboardVersion: kbVersion) {
+            try? fileManager.removeItem(at: kbPath)
+          }
+        }
+      }
+      downloadFailed(forKeyboard: request.userInfo, error: error as NSError)
+    case .downloadCachedData:
+      if request == currentRequest {
+        let error = request.error!
+        kmLog("Failed: \(error).", checkDebugPrinting: true)
+
+        currentRequest = nil
+
+        if let completionBlock = request.userInfo["completionBlock"] as? FetchKeyboardsBlock {
+          completionBlock([NSUnderlyingErrorKey: error])
+        }
+        NotificationCenter.default.post(name: .keymanLanguagesDownloadFailed, object: self,
+                                        userInfo: [NSUnderlyingErrorKey: error])
+      }
+    }
+  }
+
+  func downloadFailed(forKeyboard keyboardInfo: [AnyHashable: Any], error: NSError) {
+    let kbInfo = keyboardInfo[Key.keyboardInfo] ?? keyboardInfo
+    NotificationCenter.default.post(name: .keymanKeyboardDownloadFailed, object: self,
+                                    userInfo: [
+                                      Key.keyboardInfo: kbInfo,
+                                      NSUnderlyingErrorKey: error
+      ])
+  }
+
+  // MARK: - Loading custom keyboards
+
+  private func preloadFile(srcUrl: URL, dstDir dirUrl: URL, shouldOverwrite: Bool) {
+    let dstUrl = dirUrl.appendingPathComponent(srcUrl.lastPathComponent)
+    do {
+      if !FileManager.default.fileExists(atPath: dstUrl.path) {
+        try FileManager.default.copyItem(at: srcUrl, to: dstUrl)
+      } else if shouldOverwrite {
+        try FileManager.default.removeItem(at: dstUrl)
+        try FileManager.default.copyItem(at: srcUrl, to: dstUrl)
+      } else {
+        kmLog("File already exists at \(dstUrl) and not overwriting", checkDebugPrinting: false)
```

Comment:
Why are we not setting `checkDebugPrinting: true` here? I'm almost certain this is a logging error.

Reply:
Yeah you're right